src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -423,7 +424,11 @@ private ValueTask SendFrameLockAcquiredNonCancelableAsync(MessageOpcode opcode,
                 // the task, and we're done.
                 if (writeTask.IsCompleted)
                 {
-                    return writeTask;
+                    flushTask = _stream.FlushAsync();
```

Comment:
This is broken in few ways:
- if the write fails, the exception is being ignored if the lush completes synchronously
- if the flush completes asynchronously, the flush gets ignored
- if the flush completes asynchronously, the code issues another flush that may erroneously be concurrent with the first

Reply:
This all will be fixed with the fix you've proposed above with `.GetAwaiter().GetResult()` right? The only thing is that we might issue a second flush *after* first, in case Write completes synchronously but Flush completes asynchronously

Reply:
Yes. There could still be an unnecessary extra flush, but it'll be safe, just duplicative.

Reply:
Given that a flush on HTTP/2 will always allocate, I'd vote to avoid duplicating it if possible.

Reply:
@stephentoub do you think it's worth adding a sort of `WriteAndFlushAsync` API for consideration in #58216, or is it too niche?
On H2, the flush would be free if combined with the write. On H1, it would save the virtual call for the flush.

We also have to do this kind of flush after every write in YARP (for WebSockets & gRPC).

Reply:
> do you think it's worth adding a sort of WriteAndFlushAsync API for consideration in https://github.com/dotnet/runtime/issues/58216 (or separate issue), or is it too niche?

I think it's too niche to be worth it right now.  Adding new surface area to Stream is a big deal and needs to really be worth it.

> On H1, it would save the virtual call for the flush.

I'll be surprised if it's measurable, but even if it is, I expect it's an example where dynamic PGO will do really well to devirtualize and enable inlining it, in which case the flush will evaporate.

Reply:
I agree it's noise for H1, but for H2 it can save quite a few allocations

Reply:
> but for H2 it can save quite a few allocations

That should be solvable inside of SocketsHttpHandler if we really care.  It'd be possible to teach an Http2Stream whether all of its pending data has been flushed, in which case flush can become a non-allocating nop in this case as well.  And that would help with many more cases than this.